src/NHibernate/Impl/FutureValueAsync.cs

```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace NHibernate.Impl
+{
+	internal class FutureValueAsync<T> : IFutureValueAsync<T>, IDelayedValue
+	{
+		public delegate Task<IEnumerable<T>> GetResult();
+
+		private readonly GetResult getResult;
+
+		public FutureValueAsync(GetResult result)
+		{
+			getResult = result;
+		}
+
+		public async Task<T> GetValue()
+		{
+			var result = await getResult().ConfigureAwait(false);
```

Comment:
No cancellation token here apparently, but indeed there is one hidden in the `getResult` delegate. For me it would be more straightforward to get it as an argument of `GetValue` then transmit it to `getResult` delegate.